### DISCUSSION TITLE
useLoaderData returns null at first on client

### DISCUSSION THREAD

**Discussion by ersinakinci** [remix-run/react-router#11461] [Q&A] [answered]
I have a page nested within a layout, they look like this:

```
routes/
  _auth.login._index.tsx // page
  _auth.tsx  // layout
```

Both have `loader`s:

```ts
// auth.tsx
export const loader = async ({ request }: LoaderFunctionArgs) => {
  return 'bar';
};

export default function Layout() {
  const foo = useLoaderData<typeof loader>();

  return <Outlet />
}


// _auth.login._index.tsx
export const loader = async ({ request }: LoaderFunctionArgs) => {
  return 'abc';
};

export default function Page() {
  // Returns null on first client render
  const xyz = useLoaderData<typeof loader>();

  return ...
}
```

I'm finding that when I call `useLoaderData` on the layout route, the first client-side render of my ~page~ layout route will return `null` for `useLoaderData`. It doesn't matter what's being called or returned from either `loader`, the consistent thing is that if you ~call `useLoaderData` on the parent, the child returns `null` on the client at first.~ load a child route, the parent/layout route will return `null`.

SSR doesn't have this problem, if you stick a `console.log` in the child it will show the correct value returned from the loader on your server console. I can clearly see the loader data in the layout component's state in `window.__remixContext` as well as the HTML sent over the wire, so it's being properly serialized.

If I'm running the dev server and update a page, triggering HMR, `useLayoutData` in the layout component loads the data correctly.

My understanding is that `useLoaderData` should _always_ return whatever data has been rendered on the server. What gives?

**Comment by ersinakinci** [comment]
Argh, I was misreading my own logs. It was the _parent_ `useLoaderData` call that was returning `null`, not the child.

I think this makes sense...? Because the layout is a separate route...? So its loader doesn't get called...maybe...? I'm confused which loaders get called when. Wouldn't a layout's loader be called whenever its children are loaded, too?

Is it bad practice to put a loader in a layout?

**Comment by ersinakinci** [comment, marked as answer]
I'm not sure what I did exactly, but the problem seems to have gone away after I did some edits to `root.tsx` and did a hard refresh/clear cache on my browser.

**Comment by grinkus-adapt** [comment]
While my situation may have been slightly different, the symptoms were exactly the same. For anyone else experiencing something similar, what helped me was removing a `clientLoader` that I had set with `hydrate`. When using `hydrate` on `clientLoader` it appears that you must use `HydrateFallback` in order for it not to break loader-data in other routes. What I did is remove the `clientLoader` and fetched the required data in a component's `useEffect` (single time, on mount).
